### PR TITLE
Remove Kcas dependency

### PR DESCRIPTION
**Pull request by Sudha247**
This PR replaces `Kcas.W1` with the `Atomic` module present in the Multicore stdlib. 

Exponential backoff is now implemented with `Domain.Sync.cpu_relax`.

Results for some lockfree benchmarks:

**Time**
image

**Speedup**

image

### CODE REVIEW

**Comment by kayceesrk** [comment]
Thanks for this work @Sudha247. In the graphs, which one is 4.10.0+multicore and which one is 4.10.0+multicore+backoff+random?

**Comment by Sudha247** [comment]
In the graphs blue is 4.10.0+multicore+backoff+random, which is the current patch and 4.10.0+multicore is dotted yellow, the master branch of this repo.

File: src/lf_hash.ml [hunk 1]
```diff
@@ -107,39 +106,39 @@ module Make(Desc : HashDesc) : S = struct
...
+    let new_a = Array.init nb_bucket (fun _ -> Atomic.make Uninitialized) in
+    Atomic.set new_a.(0) (Allocated(old_access));
     let rec loop () =
-      match Cas.get t.resize with
+      match Atomic.get t.resize with
       |Some(new_access_size) as old_resize -> begin
-        if (get_size_of_access (Cas.get t.access) >= new_access_size || Cas.cas t.access old_access new_a) &&
-           ((Cas.get t.access_size) >= new_access_size || Cas.cas t.access_size old_access_size new_access_size) &&
-           ((Cas.get t.resize) <> old_resize || Cas.cas t.resize old_resize None) then
+        if (get_size_of_access (Atomic.get t.access) >= new_access_size || Atomic.compare_and_set t.access old_access new_a) &&
```
Comment (kayceesrk) [inline src/lf_hash.ml:122]: break long lines to 80 chars

File: src/lf_hash.ml [hunk 2]
```diff
@@ -107,39 +106,39 @@ module Make(Desc : HashDesc) : S = struct
...
+           ((Atomic.get t.access_size) >= new_access_size || Atomic.compare_and_set t.access_size old_access_size new_access_size) &&
```
Comment (kayceesrk) [inline src/lf_hash.ml:123]: one more long line

File: src/lf_list.ml [hunk 3]
```diff
@@ -23,9 +23,10 @@ module type S = sig
 end;;
 
 module M : S = struct
-  module Cas = Kcas.W1;;
+  (* module Cas = Kcas.W1;; *)
```
Comment (kayceesrk) [inline src/lf_list.ml:26]: delete this comment

File: src/lf_msqueue.ml [hunk 4]
```diff
@@ -34,84 +36,82 @@ end
...
-    let head = (Next (Obj.magic (), Cas.ref Nil)) in
-    { head = Cas.ref head ; tail = Cas.ref head }
+    let head = (Next (Obj.magic (), Atomic.make Nil)) in
+    { head = Atomic.make head ; tail = Atomic.make head }
 
   let is_empty q =
-    match Cas.get q.head with
-    | Nil -> failwith "MSQueue.is_empty: impossible"
+    match Atomic.get q.head with
+    | Nil -> raise Empty
```
Comment (kayceesrk) [inline src/lf_msqueue.ml:53]: I think the `failwith` is the correct behavior here. These cases cannot happen, and must be a fatal error. I'd revert this and other similar changes.

File: test/dune [hunk 5]
```diff
@@ -1,23 +1,30 @@
...
+(test
+ (name test_wsqueue_parallel)
+ (libraries lockfree unix)
+ (modules test_wsqueue_parallel))
+
+
+(test
+ (name test_msqueue_parallel)
+ (libraries lockfree unix)
+ (modules test_msqueue_parallel))
```
Comment (kayceesrk) [inline test/dune:30]: Add a newline

**Review by kayceesrk — CHANGES_REQUESTED** [review]

File: src/lf_msqueue.ml (hunk 4 above)
Comment (Sudha247) [reply src/lf_msqueue.ml]: Reverted these changes.

**Review by kayceesrk — APPROVED** [review]
lgtm

**Pull request merged**